### PR TITLE
Deepspeed Integration

### PR DESCRIPTION
**Pull request by jbloxham**
Support for using DeepSpeed instead of Pytorch DDP in the trainer. Somewhat WIP, but good enough to merge for now. The major TODOs are:

- Support for regular checkpointing
- Cleaner support for precision casting
- Config support for DeepSpeed goodies like ZeRO and activation checkpointing
- Quite a bit of testing

I'm waiting on a node to run a quick regression test.

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -224,6 +235,10 @@ def __init__(
         )
         self.eval_dl_spec = eval_dataloader_spec
 
+        # TODO: handling precision correctly with DeepSpeed is an open problem
```
Comment (Averylamp) [inline composer/trainer/trainer.py:238]: I think it'd be nice to leave these as issues as well in the code so that they can be tracked in multiple places.  
```
# TODO(issue #xyz): <explaination>
```
is nice to also be able to reference issues and make sure that comments are removed as issues are solved

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -274,13 +289,15 @@ def __init__(
...
-        if checkpoint_folder and checkpoint_interval and checkpoint_interval_unit:
+        # TODO: get checkpointing working with DeepSpeed.
+        if not self.deepspeed_enabled and checkpoint_folder and checkpoint_interval and checkpoint_interval_unit:
             self.checkpointer = Checkpointer(checkpoint_folder=get_relative_to_run_directory(checkpoint_folder),
                                              checkpoint_interval=checkpoint_interval,
                                              checkpoint_interval_unit=checkpoint_interval_unit)
 
         self.checkpoint_loader = None
-        if checkpoint_filepath:
+        # TODO: get checkpointing working with DeepSpeed.
```
Comment (Averylamp) [inline composer/trainer/trainer.py:299]: Same github issue tracking as above

**Review by Averylamp — APPROVED** [review]
LGTM, maybe wait on one more lgtm for the training loop changes

File: composer/core/precision.py [hunk 3]
```diff
@@ -8,8 +8,11 @@ class Precision(StringEnum):
     
     Attributes:
         AMP: Use :mod:`torch.cuda.amp`. Only compatible with GPUs.
+        FP16: Use 16-bit floating-point precision. Currently only
```
Comment (ravi-mosaicml) [inline composer/core/precision.py:11]: is this BF16?

File: composer/trainer/deepspeed.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+from dataclasses import dataclass
+
+import yahp as hp
+
+
+@dataclass
+class DeepSpeedHparams(hp.Hparams):
+    """Params for configuration of DeepSpeed."""
+
+    enabled: bool = hp.required("Whether to use DeepSpeed.")
```
Comment (ravi-mosaicml) [inline composer/trainer/deepspeed.py:12]: Why do we need the enabled flag? If the params are specified in the trainer, then wouldn't it be enabled?

File: composer/trainer/trainer.py [hunk 5]
```diff
@@ -174,8 +180,10 @@ def __init__(
 
         self.config = config
 
+        self.deepspeed_enabled = deepspeed_hparams and deepspeed_hparams.enabled
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:183]:
```suggestion
        self.deepspeed_enabled = deepspeed_hparams is not None
```

File: composer/trainer/deepspeed.py (hunk 4 above)
Comment (ravi-mosaicml) [reply composer/trainer/deepspeed.py:12]: Or, as an alternative, maybe just add a `use_deepspeed` flag in the trainer hparams if there are hparams?

File: composer/trainer/deepspeed.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+    def validate(self):
```
Comment (ravi-mosaicml) [inline composer/trainer/deepspeed.py:14]: Not needed if it jsut calls super

File: composer/trainer/trainer.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:238]: Can you reference a link to a stack overflow / GH issue to what is meant here (I'm a bit confused why it's an open problem)

File: composer/trainer/trainer.py [hunk 7]
```diff
@@ -224,6 +235,10 @@ def __init__(
         )
         self.eval_dl_spec = eval_dataloader_spec
 
+        # TODO: handling precision correctly with DeepSpeed is an open problem
+        precision_context = self.device.precision_context if not self.deepspeed_enabled else lambda _: contextlib.nullcontext(
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:239]:
```suggestion
        precision_context = self.device.precision_context if not self.deepspeed_enabled else contextlib.nullcontext
```

File: composer/trainer/trainer.py [hunk 8]
```diff
@@ -274,13 +289,15 @@ def __init__(
         self.state.schedulers = ComposedScheduler(schedulers=schedulers)
 
         self.checkpointer = None
-        if checkpoint_folder and checkpoint_interval and checkpoint_interval_unit:
+        # TODO: get checkpointing working with DeepSpeed.
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:292]: can you raise a `NotImplementedError` here if someone tries to use checkpointing and deepspeed?

File: composer/trainer/trainer.py [hunk 9]
```diff
@@ -9,6 +9,7 @@
 import warnings
 from typing import Any, Dict, List, Optional, Sequence, Union
 
+import deepspeed
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:12]: Should this be a conditional import? Or is deepspeed core?

File: composer/trainer/trainer.py [hunk 10]
```diff
@@ -705,15 +733,26 @@ def _train_batch_inner(self, microbatches: Sequence[Batch]):
                 for loss in ensure_tuple(state.loss):
                     loss.backward(create_graph=self.backwards_create_graph)
 
+                for loss in ensure_tuple(state.loss):
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:736]: Should this be before the loss is scaled? Or was that a bug all along?

File: tests/trainer/test_deepspeed.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+from composer.utils import ddp
+from tests.fixtures.models import SimpleBatchPairModelHparams
+from tests.trainer.test_ddp import CheckBatch0Hparams, TrackedDatasetHparams, get_batch_file_path, get_file_path
+
+
+@pytest.mark.timeout(90)
+@pytest.mark.gpu
+@pytest.mark.parametrize("world_size", [pytest.param(1), pytest.param(2, marks=pytest.mark.world_size(2))])
+def test_deepspeed(world_size: int, mosaic_trainer_hparams: TrainerHparams, tmpdir: pathlib.Path) -> None:
+    """Pretty much just copied from ./test_ddp"""
```
Comment (ravi-mosaicml) [inline tests/trainer/test_deepspeed.py:24]: #110  changes the DDP test fyi

**Review by ravi-mosaicml — COMMENTED** [review]
I'll take another look once the TODOs are adressed, but looks awesome!

File: composer/core/precision.py (hunk 3 above)
Comment (jbloxham) [reply composer/core/precision.py:11]: No. BF16 assigns different amounts of bits to the mantissa and exponent. DeepSpeed does have some BF16 support but it's very limited right now.

File: composer/trainer/deepspeed.py (hunk 4 above)
Comment (jbloxham) [reply composer/trainer/deepspeed.py:12]: When I add support for things like ZeRO, there's going to be a lot more DeepSpeed-relevant hparams, so I definitely at least want this object as a way of organizing those hparams.

File: composer/trainer/trainer.py (hunk 10 above)
Comment (jbloxham) [reply composer/trainer/trainer.py]: DeepSpeed handles loss scaling on its own, but doesn't do it in place. I've since updated the code to reflect the awkward reality that for non-DeepSpeed, we need to do this scaling before we call `loss.backward`, and for DeepSpeed, we need to scale after the backward pass.
